### ISSUE TITLE
Support for Mermaid

### ISSUE THREAD

**Issue by alaminkouser** [lumeland/lume#465]
### Enter your suggestions in details:

# Mermaid

- A graph Generator

**Comment by oscarotero** [comment]
This came up on Discord server some time ago (https://discord.com/channels/794537085641818124/1111980136691146773) but looks like it's not possible to run Mermaid on server side due https://github.com/mermaid-js/mermaid/issues/3650.

There are a couple of alternative options that you can use:

- Serea plugin: https://github.com/idris-maps/serea
- OI Lume Viz plugin: https://open-innovations.github.io/oi-lume-viz/

**Comment by oscarotero** [comment]
Yes, the only way to run mermaid is in the browser.
But it would be great to have a way to run it in build time, so there's no need to import any script in the html because the code is already generated. That's the problem with mermaid. I hope they implement server side support soon.

**Comment by alaminkouser** [comment]
Any thoughts regarding this:
https://github.com/mermaid-js/mermaid-cli

**Comment by oscarotero** [comment]
Seems that it [uses puppeteer](https://github.com/mermaid-js/mermaid-cli/blob/master/src/index.js#L404) for rendering.
I'm not sure about performance (specially if there are many diagrams and charts) but I guess we can give a try.

**Comment by alaminkouser** [comment]
deno has puppeteer. So we may avoid npm and use puppeteer in deno. I will test it.

**Comment by alaminkouser** [comment]
Here is a quick code for running puppeteer in deno and getting all svg and save the svgs in a dir:
https://github.com/alaminkouser/deno-mermaid/tree/61d526f78d71bd9ab3b1cf12f40a9b39c90e1bbc

**Comment by oscarotero** [comment]
Okay, thanks!
I'm getting this error:

```
➜  deno-mermaid git:(main) deno run -A main.ts
error: Uncaught PermissionDenied: Permission denied (os error 13)
    this.proc = Deno.run({
                     ^
    at opRun (ext:runtime/40_process.js:46:14)
    at Object.run (ext:runtime/40_process.js:132:15)
    at BrowserRunner.start (https://deno.land/x/puppeteer@16.2.0/src/deno/BrowserRunner.ts:64:22)
    at ChromeLauncher.launch (https://deno.land/x/puppeteer@16.2.0/src/deno/Launcher.ts:108:12)
    at eventLoopTick (ext:core/01_core.js:183:11)
    at async file:///Users/oscarotero/www/deno-mermaid/main.ts:3:17
```

**Comment by alaminkouser** [comment]
> Okay, thanks! I'm getting this error:
> 
> ```
> ➜  deno-mermaid git:(main) deno run -A main.ts
> error: Uncaught PermissionDenied: Permission denied (os error 13)
>     this.proc = Deno.run({
>                      ^
>     at opRun (ext:runtime/40_process.js:46:14)
>     at Object.run (ext:runtime/40_process.js:132:15)
>     at BrowserRunner.start (https://deno.land/x/puppeteer@16.2.0/src/deno/BrowserRunner.ts:64:22)
>     at ChromeLauncher.launch (https://deno.land/x/puppeteer@16.2.0/src/deno/Launcher.ts:108:12)
>     at eventLoopTick (ext:core/01_core.js:183:11)
>     at async file:///Users/oscarotero/www/deno-mermaid/main.ts:3:17
> ```

Read this: [DOC](https://deno.land/x/puppeteer@16.2.0#getting-started)

I am not sure, but you may need to run this command:
```PUPPETEER_PRODUCT=chrome deno run -A --unstable https://deno.land/x/puppeteer@16.2.0/install.ts```

**Comment by oscarotero** [comment]
@alaminkouser ok, the problem was the execution permissions for the downloaded binary.

I've been experimenting with puppeteer (Deno already supports the Node version which is more updated).
You can take a look to this experimental plugin:
https://github.com/lumeland/experimental-plugins/tree/main/puppeteer

There are some issues yet but it's a good start.

**Comment by alaminkouser** [comment]
Interesting. I will also try this.

**Comment by oscarotero** [comment]
I was able to render mermaid on server side with the Astral experimental plugin:
https://github.com/lumeland/experimental-plugins/tree/main/astral

It runs the javascript code on server side and save the generated HTML code.

For example, this page https://github.com/lumeland/experimental-plugins/blob/main/astral/demo/page.vto#L24-L28
- Run the javascript code
- Remove the script element (all elements with `astral-delete` attribute are removed)

**Comment by alaminkouser** [comment]
Amazing! Next week I will try this. It can be powerful and faster. Now generated websites can be without/less JS codes!!!

**Comment by alaminkouser** [comment]
- I cloned the [experimental repo](https://github.com/lumeland/experimental-plugins/tree/main)
- went to `astral/demo` directory
- ran `deno task build`
- It generated the website.
- It is the output:
![Screenshot from 2024-06-28 22-56-08](https://github.com/lumeland/lume/assets/36203422/8b8b8e9b-b9d0-41ba-9af2-8f959cb4a8b2)

- It is working.
- I found this output in command line while I ran `deno task build`:
![Screenshot from 2024-06-28 22-50-24](https://github.com/lumeland/lume/assets/36203422/895a3dbe-c172-4ac5-a86f-72dccb50aa0f)

It is downloading chrome. It may not use puppeteer, but uses chrome in server side.

> I tested it with a newly deno install in my computer.

**Comment by oscarotero** [comment]
Yes, it doesn't use puppeteer, but astral https://github.com/lino-levan/astral which is like Puppeteer but for Deno (Puppeteer doesn't work in Deno).

For now, the plugin is just a POC and there are some limitations (for example, it only can load assets from external urls). But for this specific use case it works pretty well.

**Comment by alaminkouser** [comment]
Only downside being it uses chrome. So cicd will be slow.

But seems stable.

**Comment by oscarotero** [comment]
I didn't play much with the library, but [seems that Firefox is also supported](https://github.com/lino-levan/astral/blob/302a2aa858b3bc5dfc54758cfdc27f10acfd658b/src/browser.ts#L69).

If no browser is specified, it downloads chrome automatically.

**Comment by ooker777** [comment]
Discussion for anyone interested: [What makes mermaid.js cannot work on server-side, while other visualization libs can?](https://github.com/lumeland/lume/discussions/641 "What makes mermaid.js cannot work on server-side, while other visualization libs can? · lumeland/lume · Discussion #641")

**Comment by ooker777** [comment]
FYI I made a plugin to inject Mermaid script at the front end: https://github.com/ooker777/lume-mermaid-plugin

**Comment by foster-hangdaan** [comment]
To add to the list of options, I've created a [Kroki plugin for Lume](https://code.fosterhangdaan.com/foster/lume-plugin-kroki) which converts codeblocks containing diagram descriptions into an `<img/>` during the site build. It currently has support for all of the [Kroki Diagram Types](https://kroki.io/#support); Mermaid being one of them.

The images are rendered by a [Kroki](https://kroki.io/) server. By default, the plugin uses the official instance at https://kroki.io, but can be configured to use a self-hosted instance.
